File: behavioral/visitor/code_analyzer/typescript/code_analyzer/visitor.ts
import {
  FunctionDefinitionNode,
  VariableDeclarationNode,
  AssignmentStatementNode,
  IfStatementNode,
  ExpressionStatementNode,
} from "./ast_nodes"; // Import concrete node types

// Visitor interface declaring visit methods for each concrete element type.
export interface Visitor {
  visitFunctionDefinition(element: FunctionDefinitionNode): void;
  visitVariableDeclaration(element: VariableDeclarationNode): void;
  visitAssignmentStatement(element: AssignmentStatementNode): void;
  visitIfStatement(element: IfStatementNode): void;
  visitExpressionStatement(element: ExpressionStatementNode): void;
}

File: behavioral/visitor/code_analyzer/go/code_analyzer/visitor.go
package code_analyzer

// Visitor interface declares visit methods for each concrete element type.
// Using interface{} for element allows different node types, but requires type assertion in implementations.
type Visitor interface {
	VisitFunctionDefinition(element *FunctionDefinitionNode)
	VisitVariableDeclaration(element *VariableDeclarationNode)
	VisitAssignmentStatement(element *AssignmentStatementNode)
	VisitIfStatement(element *IfStatementNode)
	VisitExpressionStatement(element *ExpressionStatementNode)
}
